src/pcp2openmetrics/pcp2openmetrics.py

```diff
@@ -0,0 +1,569 @@
+        if self.runtime != -1:
+            sys.stdout.write(':" }\n{ "//": "%s samples(s) with %.1f sec interval ~ %d sec runtime." }\n' % (self.samples, float(self.interval), self.runtime))
+        elif self.samples:
+            duration = (self.samples - 1) * float(self.interval)
+            sys.stdout.write(':" }\n{ "//": "%s samples(s) with %.1f sec interval ~ %d sec runtime." }\n' % (self.samples, float(self.interval), duration))
+        else:
+            sys.stdout.write('..." }\n{ "//": "(Ctrl-C to stop)" }\n')
+
+    def write_openmetrics(self, timestamp):
+        """ Write results in openmetrics format """
+        if timestamp is None:
+            # Silent goodbye, close in finalize()
+            return
+
+        ts = self.context.datetime_to_secs(self.pmfg_ts(), PM_TIME_SEC)
+
+        if self.prev_ts is None:
+            self.prev_ts = ts
+
+        if not self.writer and not self.url:
+            if self.outfile is None:
+                self.writer = sys.stdout
+            else:
+                self.writer = open(self.outfile, 'wt')
+
+        def get_type_string(desc):
+            """ Get metric type as string """
+            if desc.contents.type == pmapi.c_api.PM_TYPE_32:
+                mtype = "32-bit signed"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_U32:
+                mtype = "32-bit unsigned"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_64:
+                mtype = "64-bit signed"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_U64:
+                mtype = "64-bit unsigned"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_FLOAT:
+                mtype = "32-bit float"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_DOUBLE:
+                mtype = "64-bit float"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_STRING:
+                mtype = "string"
+            else:
+                mtype = "unknown"
+            return mtype
+
+        def openmetrics_name(metric):
+            """ pcp.io metric name to openmetrics.io name conventions """
+            return metric.replace('.','_')
+        
+        def openmetrics_help(context, pmid):
+            """ convert pcp.io metric helptext to openmetrics.io HELP """
+            return context.pmLookupText(pmid[0]).decode()
+        
+        def openmetrics_type(context, pmid):
+            """ convert pcp.io metric metadata to openmetrics.io TYPE """
+            desc = context.pmLookupDesc(pmid[0])
+            if desc.sem == cpmapi.PM_SEM_COUNTER:
+                return 'counter'
+            return 'gauge'
+        
+        def openmetrics_labels(metric, context, inst, name, desc):
+            """ filter pcp.io labels here; pick out the labels needed """
+            result = ''
+            pmid = context.pmLookupName(metric)
+            labels = context.pmLookupLabels(pmid[0])
```

Comment:
Both of these calls involve round trips communication to pmcd - we already have the pmid in the desc fortunately.  The labels are like the help text - we should do a single lookup at the start of the program for all metrics, store it in a dictionary, and do a quick lookup here.